### PR TITLE
Add flag to disable MPI threads - HOROVOD_MPI_THREADS_DISABLE

### PR DESCRIPTION
**Pull request by alsrgv**

### CODE REVIEW

File: horovod/common/operations.cc [hunk 1]
```diff
@@ -1179,10 +1179,21 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
...
   //
-  // We will ask for multiple threads, so other libraries like mpi4py can
+  // In some cases MPI library has multi-threading support, but it slows down certain
+  // components, e.g. OpenIB BTL in OpenMPI gets disabled if MPI_THREAD_MULTIPLE is
+  // requested.
+  //
+  // By default, we will ask for multiple threads, so other libraries like mpi4py can
   // be used together with Horovod if multi-threaded MPI is installed.
+  auto mpi_threads_disable = std::getenv("HOROVOD_MPI_THREADS_DISABLE");
+  int required;
```
Comment (jiezhang) [inline horovod/common/operations.cc]: This seems cleaner.
```
int required = MPI_THREAD_MULTIPLE;
if (mpi_threads_disable != nullptr && std::atoi(mpi_threads_disable) > 0) {
  required = MPI_THREAD_FUNNELED;
}
```

Comment (jeevandev) [reply horovod/common/operations.cc]: 2nd Jie's comment

File: docs/running.md [hunk 2]
```diff
@@ -12,7 +12,9 @@ allows you to have a mixture of different NUMA configurations because the defaul
 `-mca pml ob1` and `-mca btl ^openib` flags force the use of TCP for MPI communication.  This avoids many multiprocessing
 issues that Open MPI has with RDMA which typically result in segmentation faults.  Using TCP for MPI does not have
 noticeable performance impact since most of the heavy communication is done by NCCL, which will use RDMA via RoCE or
-InfiniBand if they're available (see [Horovod on GPU](gpus.md)).
+InfiniBand if they're available (see [Horovod on GPU](gpus.md)).  Notable exception from this rule are models that heavily
```
Comment (sblotner) [inline docs/running.md]: exception --> exceptions

File: docs/running.md [hunk 3]
```diff
@@ -12,7 +12,9 @@ allows you to have a mixture of different NUMA configurations because the defaul
...
+use `hvd.broadcast()` and `hvd.allgather()` operations.  To make those operations use RDMA read [Open MPI with RDMA](#open-mpi-with-rdma)
```
Comment (sblotner) [inline docs/running.md]: "RDMA read [Open MPI...]" --> "RDMA, read the [Open MPI...]"

File: docs/running.md [hunk 4]
```diff
@@ -39,6 +41,25 @@ $ mpirun -np 16 \
     python train.py
 ```
 
+### Open MPI with RDMA
+
+As noted above, using TCP for MPI communication does not have any significant affects on performance in the majority of cases.
```
Comment (sblotner) [inline docs/running.md]: affects --> effects

File: docs/running.md [hunk 5]
```diff
@@ -39,6 +41,25 @@ $ mpirun -np 16 \
...
+Models that make heavy use of `hvd.broadcast()` and `hvd.allgather()` operations are exceptions to that rule.
+
+Default Open MPI `openib` BTL that provides RDMA functionality does not work well with MPI multi-threading.  In order to use
+RDMA with `openib`, multi-threading must be disabled via `-x HOROVOD_MPI_THREADS_DISABLE=1` option.  See example below:
```
Comment (sblotner) [inline docs/running.md]: See example --> See the example

File: docs/running.md [hunk 6]
```diff
@@ -39,6 +41,25 @@ $ mpirun -np 16 \
...
+```bash
+$ mpirun -np 16 \
+    -H server1:4,server2:4,server3:4,server4:4 \
+    -bind-to none -map-by slot \
+    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x HOROVOD_MPI_THREADS_DISABLE=1 -x PATH \
+    -mca pml ob1 \
+    python train.py
+```
+
+Other MPI RDMA implementations may or may not benefit from disabling multi-threading, please consult vendor documentation.
```
Comment (sblotner) [inline docs/running.md]: multi-threading, please --> multithreading, so please

**Pull request merged**